tests/tests/relayer.rs

```diff
@@ -477,3 +484,196 @@ async fn handle(
 }
+
+#[tokio::test(flavor = "multi_thread")]
+async fn balances_do_not_return_retryable_messages() {
+    let mut rng = StdRng::seed_from_u64(1234);
+    let mut config = Config::local_node();
+    config.relayer = Some(relayer::Config::default());
+    let relayer_config = config.relayer.as_mut().expect("Expected relayer config");
+    let eth_node = MockMiddleware::default();
+    let contract_address = relayer_config.eth_v2_listening_contracts[0];
+    const TIMEOUT: Duration = Duration::from_secs(1);
+
+    // Large enough to get all messages, but not to trigger the "query is too complex" error.
+    const UNLIMITED_QUERY: i32 = 100;
+
+    // Given
+
+    // setup a retryable and non-retryable message
+    let secret_key: SecretKey = SecretKey::random(&mut rng);
+    let public_key = secret_key.public_key();
+    let recipient = Input::owner(&public_key);
+
+    const RETRYABLE_AMOUNT: u64 = 99;
+    const RETRYABLE_NONCE: u64 = 0;
+    const NON_RETRYABLE_AMOUNT: u64 = 100;
+    const NON_RETRYABLE_NONCE: u64 = 1;
+    let messages = vec![
+        MessageKind::Retryable {
+            nonce: RETRYABLE_NONCE,
+            amount: RETRYABLE_AMOUNT,
+        },
+        MessageKind::NonRetryable {
+            nonce: NON_RETRYABLE_NONCE,
+            amount: NON_RETRYABLE_AMOUNT,
+        },
+    ];
+    let logs: Vec<_> = setup_messages(&messages, &recipient, &contract_address);
+
+    eth_node.update_data(|data| data.logs_batch = vec![logs.clone()]);
+    // Setup the eth node with a block high enough that there
+    // will be some finalized blocks.
+    eth_node.update_data(|data| data.best_block.number = Some(200.into()));
+    let eth_node = Arc::new(eth_node);
+    let eth_node_handle = spawn_eth_node(eth_node).await;
+
+    relayer_config.relayer = Some(vec![format!("http://{}", eth_node_handle.address)
+        .as_str()
+        .try_into()
+        .unwrap()]);
+
+    config.utxo_validation = true;
+
+    // setup fuel node with mocked eth url
+    let db = Database::in_memory();
+
+    let srv = FuelService::from_database(db.clone(), config)
+        .await
+        .unwrap();
+
+    let client = FuelClient::from(srv.bound_address);
+    let base_asset_id = client
+        .consensus_parameters(0)
+        .await
+        .unwrap()
+        .unwrap()
+        .base_asset_id()
+        .clone();
+
+    // When
+
+    // wait for relayer to catch up to eth node
+    srv.await_relayer_synced().await.unwrap();
+    // Wait for the block producer to create a block that targets the latest da height.
+    srv.shared
+        .poa_adapter
+        .manually_produce_blocks(
+            None,
+            Mode::Blocks {
+                number_of_blocks: 1,
+            },
+        )
+        .await
+        .unwrap();
+
+    // Balances are processed in the off-chain worker, so we need to wait for it
+    // to process the messages before we can assert the balances.
+    let result = tokio::time::timeout(TIMEOUT, async {
+        loop {
```

Comment:
Is there a better way to ensure that off chain worker has processed the events?

Reply:
I don't think so. Ideally we should be able to register and listen for events from the tracing module.
To me it looks like waiting for the balances to be available is a viable alternative